tools/cfn2ts/lib/codegen.ts

```diff
@@ -593,20 +593,71 @@ export default class CodeGenerator {
+          .join(', ');
+
+        return `${CFN_PARSE}.FromCloudFormation.getArray(${this.baseExpression}, ` +
+          `(${arg1}: any) => ${CFN_PARSE}.FromCloudFormation.getTypeUnion([${validatorNames}], [${mappers}], ${arg1})` +
+          ')';
       }
 
-      public visitUnionList(_itemTypes: genspec.CodeName[]): string {
-        return this.baseExpression;
+      public visitUnionMap(itemTypes: genspec.CodeName[]): string {
+        const validatorNames = itemTypes.map(type => genspec.validatorName(type).fqn).join(', ');
+
+        const arg1 = `prop${this.depth}`;
+        const arg2 = `prop${this.depth + 1}`;
+        const mappers = itemTypes
+          // we need a copy 2 levels deep, as we're already using prop2 in arg2 here
+          .map(type => `(${arg2}: any) => ${this.deeperCopy(arg1).deeperCopy(arg2).visitAtom(type)}`)
+          .join(', ');
+
+        return `${CFN_PARSE}.FromCloudFormation.getMap(${this.baseExpression}, ` +
+          `(${arg1}: any) => ${CFN_PARSE}.FromCloudFormation.getTypeUnion([${validatorNames}], [${mappers}], ${arg1})` +
+          ')';
       }
 
-      public visitUnionMap(_itemTypes: genspec.CodeName[]): string {
-        return this.baseExpression;
+      public visitListOrAtom(scalarTypes: genspec.CodeName[], itemTypes: genspec.CodeName[]): any {
+        const scalarValidatorNames = scalarTypes.map(type => genspec.validatorName(type).fqn).join(', ');
+        const itemValidatorNames = itemTypes.map(type => genspec.validatorName(type).fqn).join(', ');
+
+        const arg1 = `prop${this.depth}`;
+        const scalarTypesMappers = scalarTypes
+          .map(type => `(${arg1}: any) => ${this.deeperCopy(arg1).visitAtom(type)}`)
+          .join(', ');
+        const scalarMapper = `(${arg1}: any) => ${CFN_PARSE}.FromCloudFormation.getTypeUnion([${scalarValidatorNames}], [${scalarTypesMappers}], ${arg1})`;
+
+        const arg2 = `prop${this.depth + 1}`;
+        const arg3 = `prop${this.depth + 2}`;
+        const itemTypeMappers = itemTypes
+          // we need a copy 3 levels deep, because we're using prop3 already in arg3
+          .map(type => `(${arg3}: any) => ${this.deeperCopy(arg1).deeperCopy(arg2).deeperCopy(arg3).visitAtom(type)}`)
+          .join(', ');
+        const listMapper = `(${arg1}: any) => ${CFN_PARSE}.FromCloudFormation.getArray(${arg1}, ` +
+          `(${arg2}: any) => ${CFN_PARSE}.FromCloudFormation.getTypeUnion([${itemValidatorNames}], [${itemTypeMappers}], ${arg2})` +
+          ')';
+
+        const scalarValidator = `${CORE}.unionValidator(${scalarValidatorNames})`;
+        const listValidator = `${CORE}.listValidator(${CORE}.unionValidator(${itemValidatorNames}))`;
+
+        return `${CFN_PARSE}.FromCloudFormation.getTypeUnion([${scalarValidator}, ${listValidator}], [${scalarMapper}, ${listMapper}], ${this.baseExpression})`;
       }
 
       private deeperCopy(baseExpression: string): FromCloudFormationFactoryVisitor {
```

Comment:
Once we get rid of the depth, this doesn't really need any instance variables anymore, and the "copy" word implies a lot of statekeeping that is making me uncomfortable. Wouldn't factory functions make more sense?

```ts
FromCloudFormationFactoryVisitor.fromExpression(...)
// ^--- this is a little long then, might want to rename it
```

Reply:
Can't get rid of `depth` because of https://github.com/aws/aws-cdk/pull/9442#discussion_r466023824